Document: Fix - User can send unsafe settings.

matipojo opened the pull request.

matipojo committed:
Merge remote-tracking branch 'elementor/develop' into ED-2229-control-select-options-xss

matipojo committed:
Document: Fix - User can send unsafe settings.

The setting for controls from type Select/Select2 was not validated against the original options., in some elements, it was possible to inject JavaScript code divided into few controls, in a way that can't be escaped by the `wp_kses` function.

ref: ED-2229

rodolphebertozzo commented on core/base/document.php:
```diff
@@ -1375,4 +1390,58 @@ private function add_handle_revisions_changed_filter() {
+	 */
+	private function validate_controls( array $settings, array $controls ) {
+		foreach ( $controls as $control_id => $control_config ) {
+			if ( Controls_Manager::SELECT !== $control_config['type'] && Controls_Manager::SELECT2  !== $control_config['type'] ) {
```
```suggestion
			if ( Controls_Manager::SELECT !== $control_config['type'] && Controls_Manager::SELECT2 !== $control_config['type'] ) {
```

matipojo committed:
Document: Fix - PHP Lint.

kobizz commented on core/base/document.php:
```diff
@@ -1375,4 +1390,59 @@ private function add_handle_revisions_changed_filter() {
+	 *
+	 * @return array
+	 */
+	private function validate_elements_settings( array $elements ) {
```
Please make it a general loop so we can use it for other purposes (such as saving assets used by the page)

kobizz commented on core/base/document.php:
```diff
@@ -1375,4 +1390,59 @@ private function add_handle_revisions_changed_filter() {
+	 *
+	 * @return array
+	 */
+	private function validate_controls( array $settings, array $controls ) {
```
It's better to add a method inside the control class for validating saved values

matipojo committed:
Document: Fix - Save validations resets some control values.

Some cases were not handled. like `groups` on a` select` control, and `multiple` on a `select2` control. also added a `before_save` method to handle in the future more things besides the validations.

matipojo committed:
Document: Fix - PHPUnit fails with a wrong typing.
